manual/src/main/scala/pl/metastack/metarouter/manual/Examples.scala

```diff
@@ -20,42 +22,27 @@ object Examples extends SectionSupport {
   }
 
   section("map") {
-    case class Details(userId: Int)
-    val details = Router.route[Details](Root / "details" / Arg[Int])
     Router.parse(details, "/details/42")
   }
 
   section("parse") {
-    case class Details(userId: Int)
-    case class UserInfo(user: String, details: Boolean)
-
-    val details  = Router.route[Details](Root / "details" / Arg[Int])
-    val userInfo = Router.route[UserInfo](Root / "user" / Arg[String] / Arg[Boolean])
-
-    val routes = Router.create(details).orElse(userInfo)
-    routes.parse("/user/hello/false")
-  }
-
-  case class Details(userId: Int)
-  case class UserInfo(user: String, details: Boolean)
-
-  section("urls") {
-    val details  = Router.route[Details](Root / "details" / Arg[Int])
-    val userInfo = Router.route[UserInfo](Root / "user" / Arg[String] / Arg[Boolean])
+    val details  = Root / "details" / Arg[Int]
+    val userInfo = Root / "user" / Arg[String] / Arg[Boolean]
 
-    List(
-      Router.url(details, Details(42)),
-      Router.url(userInfo, UserInfo("test", true))
-    )
+    "/user/hello/false" match {
```

Comment:
This is a very compact and elegant solution! Even though Shapeless is geared towards compile-time and `match` is runtime, this solution may well be optimal - what belongs to runtime, must stay at runtime.

Implementing this with Shapeless will most likely involve coproducts and Poly. Something like:

```scala
val parsed: Error :+: details.type :+: userInfo.type = "/user/hello/false".parse  // via rich wrappers
parsed.fold {  // or not fold, I don't know for sure
  implicit val at[details.type] = ...
  implicit val at[userInfo.type] = ...
}
```

The marginal gain from purely compile-time solution is that it will force you to consider the case of error. But you'll end up writing much more code. So I think your current solution is optimal.